big_tests/tests/smart_markers_SUITE.erl

```diff
@@ -164,27 +319,94 @@ delete_room(Owner, Users, RoomBinJid) ->
         MsgId = escalus_stanza:id(),
-        Msg = escalus_stanza:set_id(
-                escalus_stanza:groupchat_to(RoomBinJid, <<"Hello">>), MsgId),
+        StzId = send_msg_to_room(Users, RoomBinJid, Writer, MsgId),
+        mark_msg(Users, RoomBinJid, Marker, StzId).
+
+send_msg_to_room(Users, RoomBinJid, Writer, MsgId) ->
+        Msg = escalus_stanza:set_id(escalus_stanza:groupchat_to(RoomBinJid, <<"Hello">>), MsgId),
         escalus:send(Writer, Msg),
-        [ escalus:wait_for_stanza(User) || User <- Users],
+        Msgs = [ escalus:wait_for_stanza(User) || User <- Users],
+        get_id(hd(Msgs), MsgId).
+
+mark_msg(Users, RoomBinJid, Marker, StzId) ->
         ChatMarker = escalus_stanza:setattr(
-                       escalus_stanza:chat_marker(RoomBinJid, <<"displayed">>, MsgId),
+                       escalus_stanza:chat_marker(RoomBinJid, <<"displayed">>, StzId),
                        <<"type">>, <<"groupchat">>),
         escalus:send(Marker, ChatMarker),
         [ escalus:wait_for_stanza(User) || User <- Users],
-        MsgId.
+        StzId.
 
 send_message_respond_marker(MsgWriter, MarkerAnswerer) ->
-    Body = <<"Hello">>,
+    send_message_respond_marker(MsgWriter, MarkerAnswerer, undefined).
+
+send_message_respond_marker(MsgWriter, MarkerAnswerer, MaybeThread) ->
     MsgId = escalus_stanza:id(),
-    Msg = escalus_stanza:set_id(escalus_stanza:chat_to(MarkerAnswerer, Body), MsgId),
+    Msg = add_thread_id(escalus_stanza:set_id(
+                          escalus_stanza:chat_to(MarkerAnswerer, <<"Hello!">>),
+                          MsgId),
+                        MaybeThread),
     escalus:send(MsgWriter, Msg),
     escalus:wait_for_stanza(MarkerAnswerer),
-    ChatMarker = escalus_stanza:chat_marker(MsgWriter, <<"displayed">>, MsgId),
+    ChatMarker = add_thread_id(escalus_stanza:chat_marker(
+                                 MsgWriter, <<"displayed">>, MsgId),
+                               MaybeThread),
     escalus:send(MarkerAnswerer, ChatMarker),
     escalus:wait_for_stanza(MsgWriter).
 
+verify_marker_fetch(MarkingUser, MarkedUser) ->
+    verify_marker_fetch(MarkingUser, MarkedUser, undefined, undefined).
+
+verify_marker_fetch(MarkingUser, MarkedUser, Thread, After) ->
+        MarkedUserBJid = case is_binary(MarkedUser) of
+                             true -> [{<<"peer">>, MarkedUser}];
+                             false -> [{<<"peer">>, escalus_utils:jid_to_lower(escalus_client:short_jid(MarkedUser))}]
+                         end,
+        MaybeThread = case Thread of
+                          undefined -> [];
+                          _ -> [{<<"thread">>, Thread}]
+                      end,
+        MaybeAfter = case After of
+                          undefined -> [];
+                          _ -> [{<<"after">>, After}]
+                      end,
+        Iq = iq_fetch_marker(MarkedUserBJid ++ MaybeThread ++ MaybeAfter),
+        mongoose_helper:wait_until(
```

Comment:
wait until which waits for what exactly? (probably for not a crash, but still)

Reply:
Until it all returns fine, without crashing. There's a bunch of asserts inside that might fail but are eventually working, it just needs to fetch the iq again.